src/rp2/plugin/accounting_method/hifo.py

```diff
@@ -172,20 +172,35 @@ def _seek_first_non_exhausted_acquired_lot(self, acquired_lot_list: List[InTrans
         # This while loop causes O(nm) complexity, where n is the number of taxable events and m is the number of acquired lots):
         # for every taxable event, loop over the acquired lot list. There are non-trivial ways of making this faster (by changing
         # the data structures).
+        acquired_lot_amount: RP2Decimal = ZERO
+        acquired_lot: Optional[InTransaction] = None
         for index in range(start, -1, -1):
-            acquired_lot: InTransaction = acquired_lot_list[index]
-            acquired_lot_amount: RP2Decimal
+            # if the next next is less than us or zero partial, skip
+            if (
+                acquired_lot
+                and acquired_lot.spot_price > acquired_lot_list[index].spot_price
+                or (self._has_partial_amount(acquired_lot_list[index]) and self._get_partial_amount(acquired_lot_list[index]) <= ZERO)
+            ):
+                continue
+            else:
+                acquired_lot_amount = ZERO
+                acquired_lot = acquired_lot_list[index]
             if self._has_partial_amount(acquired_lot):
                 if self._get_partial_amount(acquired_lot) > ZERO:
                     acquired_lot_amount = self._get_partial_amount(acquired_lot)
-                    self._clear_partial_amount(acquired_lot)
-                    return AcquiredLotAndAmount(acquired_lot=acquired_lot, amount=acquired_lot_amount)
+                else:
+                    acquired_lot = None  # if we ourselves are zero, clear this variable so the next itteration forces picking out of the array
+                    # edge case, but if the first one is a depleted lot, return none
+                    if index == start:
```

Comment:
Not sure I get this case: could you expand on it? In the first iteration it will always be index == start.

Reply:
I'm glad you asked. I'm wondering if I'm working around what might be a different issue in the calling method `get_acquired_lot_for_taxable_event`. My initial change worked for where I was seeing the lots diverge from my golden copy. Later on in same input though they diverged again. In this divergence though, the lot selected was one of lesser cost basis than what was available.

What seemed to be happening was `_seek_first_non_exhausted_acquired_lot` was getting called with a start index lower than what should have been tried, as there were InTransactions at higher indexes of `acquired_lot_list` which were still acquired before the taxable event date. I got around it by returning None when the very first lot examined had 0 partial. This forced the method to get called again but with a higher start index. My workaround might have been very specific to my input. I think I'll try coming up with some example input or try using the examples in the repo to demonstrate.

Reply:
Hmm, interesting... As you said, it would be great if we can get a smallish repro (like test_data2 for the initial problem). The caller is getting lots from the AVLTree that is setup in `initialize()`. The AVLTree organizes lots by spot_price, with some interesting disambiguation logic explained here: https://github.com/eprbell/rp2/blob/main/src/rp2/plugin/accounting_method/hifo.py#L76. I'll also review the code again to see if I can catch a problem with the logic.

Reply:
I managed to make [a small test data set](https://github.com/debben/rp2/commit/b60afe9cfb2839bb8713998c008d687af6d31025) in another branch which demonstrates both #73 as well as what I worked around here.

Without any changes to hifo, the issue #73 shows for the 1/23 out transaction, where the second lot should have been the remaining 20% of the 1/8 lot. Instead the 1/18 lot is selected because it is the first non-zero partial found.

With my patch (minus handling of that edge case on 194), the lots are selected correctly until the 1/25 out transaction, where the first lot should have been the one acquired 1/24 at a spot price of 39. Instead the last fraction of the 1/8 lot (spot price of 28) is used.

Reply:
Thanks for creating a repro! I'll take a look at it.